clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,952 @@
+            log.debug("initiate close for client telemetry, check if terminal push required. Timeout {} ms.", timeoutMs);
+
+            lock.writeLock().lock();
+            try {
+                // If we never fetched a subscription, we can't really push anything.
+                if (lastRequestMs == 0) {
+                    log.debug("Telemetry subscription not loaded, not attempting terminating push");
+                    return;
+                }
+
+                if (state == ClientTelemetryState.SUBSCRIPTION_NEEDED) {
+                    log.debug("Subscription not yet loaded, ignoring terminal push");
+                    return;
+                }
+
+                if (isTerminatingState() || !maybeSetState(ClientTelemetryState.TERMINATING_PUSH_NEEDED)) {
+                    log.debug("Ignoring subsequent initiateClose");
+                    return;
+                }
+
+                try {
+                    log.info("About to wait {} ms. for terminal telemetry push to be submitted", timeoutMs);
+                    if (!terminalPushInProgress.await(timeoutMs, TimeUnit.MILLISECONDS)) {
+                        log.info("Wait for terminal telemetry push to be submitted has elapsed; may not have actually sent request");
+                    }
+                } catch (InterruptedException e) {
+                    log.warn("Error during client telemetry close", e);
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+
+        private Optional<Builder<?>> createSubscriptionRequest(ClientTelemetrySubscription localSubscription) {
+            // If we've previously retrieved a subscription, it will contain the client instance ID
+            // that the broker assigned. Otherwise, per KIP-714, we send a special "null" UUID to
+            // signal to the broker that we need to have a client instance ID assigned.
+            Uuid clientInstanceId = (localSubscription != null) ? localSubscription.clientInstanceId() : Uuid.ZERO_UUID;
+
+            lock.writeLock().lock();
+            try {
+                if (isTerminatingState()) {
+                    return Optional.empty();
+                }
+
+                if (!maybeSetState(ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS)) {
+                    return Optional.empty();
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            AbstractRequest.Builder<?> requestBuilder = new GetTelemetrySubscriptionsRequest.Builder(
+                new GetTelemetrySubscriptionsRequestData().setClientInstanceId(clientInstanceId), true);
+            return Optional.of(requestBuilder);
+        }
+
+        private Optional<Builder<?>> createPushRequest(ClientTelemetrySubscription localSubscription) {
+            if (localSubscription == null) {
+                log.warn("Telemetry state is {} but subscription is null; not sending telemetry", state);
```

Comment:
maybe move this below the maybeSetState.  The order makes a bit more sense for me: I was thinking the logical order would be:

1. Unable to transition....
2. [therefore]... not sending telemetry

As you are logging at error level - it seems like we aren't doing anything after a failed state transition. Should we reset the state or something? Would it cause repeated error messages being logged?  If this is not a fatal condition, maybe we could just warn?

Reply:
Thanks @philipnee. I think the order is correct as I do want to log the state prior re-setting the state from `mayBeSetState`. 

I agree that the log level shouldn't be `error` and I changed to `warn` if state cannot be updated. If state transition is not successful, which should not be the case ideally, then the error log might appear after push interval is elapsed.

I did think about re-setting to some initial `state` in case of error but if the client state transition is buggy then the problem will re-occur. I have added line to disable telemetry in case of error state transition.